src/parser/SparqlLexer.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright 2019, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@neptun.uni-freiburg.de)
```

Comment:
One thing that is different with this compared to a classic generated lexer that constructs FSAs over all token definitions, is that you need to handle it manuallly if at one point in the parser a token can match the prefix of another token. In a traditional lexer (as I understand it) the lexer would keep trying to extend the token and then only emit the longest match. I haven't found any situation in our SPARQL parser where this would be a problem, but I think we should be aware of this and we should add a comment so anyone coming after us is also aware of this quirk.

Reply:
I've added a comment into the the `readNext` method, but I think the code should be quite self explanatory as to the way multiple matches are handled (given that the core function that extracts tokens is mostly a lot of `else if`)